### PR TITLE
Refactor SyncJob class

### PR DESCRIPTION
**Pull request by wangch079**
This PR refactors `SyncJob` class. The changes are:

1. Add a new class `ESDocument`, to be a parent class of `SyncJob` and `Connector` class.
2. Make `SyncJob` data representation of a document in `.elastic-connectors-sync-jobs`, and include self-mutating methods.
3. Move `transform_filtering` method from `SyncJob` to `Filter`.
4. Add `validate_filtering` method to `SyncJob`.
5. Make the system work with the refactored `SyncJob`.

#### Pre-Review Checklist
- [x] Covered the changes with automated tests
- [x] Tested the changes locally by running `make test` and `make ftest NAME=mysql`
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [ ] Considered corresponding documentation changes
- [ ] Contributed any configuration settings changes to the configuration reference

### CODE REVIEW

File: connectors/byoc.py [hunk 1]
```diff
@@ -178,96 +178,142 @@ async def all_connectors(self):
...
-            if (filtering is None or len(filtering) == 0)
-            else deepcopy(filtering)
+    async def validate_filtering(self, validator):
+        validation_result = await validator.validate_filtering(self.filtering)
+
+        if validation_result.state != FilteringValidationState.VALID:
+            raise InvalidFilteringError(
+                f"Filtering in state {validation_result.state}. Expected: {FilteringValidationState.VALID}."
+            )
+        if len(validation_result.errors):
```
Comment (wangch079) [inline connectors/byoc.py:267]: Hi @timgrein , is there any scenario where the state is `VALID` but `errors` is not empty?

File: connectors/byoc.py [hunk 2]
```diff
@@ -321,6 +355,19 @@ def has_advanced_rules(self):
...
+        """
+        Transform the filtering in .elastic-connectors to filtering ready-to-use in .elastic-connectors-sync-jobs
+        """
+        # deepcopy to not change the reference resulting in changing .elastic-connectors filtering
+        filtering = (
+            {"advanced_snippet": {}, "rules": []} if len(self) == 0 else deepcopy(self)
+        )
+
+        # extract value for sync job
+        filtering["advanced_snippet"] = self.advanced_rules
```
[the remark this answers is not in the document]
Comment (timgrein) [reply connectors/byoc.py:380]: If you check the protocol you'll see that `advanced_snippet` nests everything under `value`, when it stores it alongside rules in the `.elastic-connectors`. As a sync job does not nest the actual content under `value` this is expected.

So the original filtering will never be: 
```
{
  "advanced_snippet": {
     "query": {}
     }, 
   "rules": []
}
```
but would be 
```
{
  "advanced_snippet": {
     "value": {
       "query": {}
      }, 
     "rules": []
}
```.

So yes this is expected. Did you test it locally?

File: connectors/byoc.py (hunk 1 above)
Comment (timgrein) [reply connectors/byoc.py:267]: No scenario from our side, but I want to be as robust as possible against connector implementations outside of our control. We could use an invariant (similar to "Design by Contract") and add an `assert` here, which simply signals a connector developer, that the provided filtering implementation is semantically wrong (or we are not as strict and log a warning or an error).

File: connectors/byoc.py [hunk 3]
```diff
@@ -178,96 +178,142 @@ async def all_connectors(self):
...
-        self.job_id = resp["_id"]
-        return self.job_id
+    @property
+    def configuration(self):
+        return DataSourceConfiguration(self.get("connector", "configuration"))
 
-    async def done(self, indexed_count=0, deleted_count=0, exception=None):
-        self.completed_at = datetime.now(timezone.utc)
+    @property
+    def filtering(self):
```
Comment (timgrein) [inline connectors/byoc.py:233]: Shouldn't `filtering` be of type array in a sync job? (see: [Protocol](https://github.com/elastic/connectors-ruby/blob/main/docs/CONNECTOR_PROTOCOL.md)) Btw I think we need to update these docs and maybe add them also to connectors-python.

File: connectors/byoc.py [hunk 4]
```diff
@@ -178,96 +178,142 @@ async def all_connectors(self):
...
+
+    async def done(self, ingestion_stats={}, connector_metadata={}):
+        await self._terminate(
+            JobStatus.COMPLETED, None, ingestion_stats, connector_metadata
         )
 
-        # extract value for sync job
-        filtering["advanced_snippet"] = filtering.get("advanced_snippet", {}).get(
-            "value", {}
+    async def fail(self, message, ingestion_stats={}, connector_metadata={}):
```
Comment (timgrein) [inline connectors/byoc.py:286]: Should we call this `error` as the names of `cancel` and `suspend` also adhere to the corresponding `JobStatus`?

File: connectors/byoc.py [hunk 5]
```diff
@@ -178,96 +178,142 @@ async def all_connectors(self):
...
+    async def claim(self):
+        doc = {
+            "status": e2str(JobStatus.IN_PROGRESS),
+            "started_at": iso_utc(),
+            "last_seen": iso_utc(),
+            "worker_hostname": socket.gethostname(),
+        }
+        await self.index.update(doc_id=self.id, doc=doc)
+
+    async def done(self, ingestion_stats={}, connector_metadata={}):
```
Comment (timgrein) [inline connectors/byoc.py:281]: Should we call this `complete` as the names of `cancel` and `suspend` also adhere to the corresponding `JobStatus`?

File: connectors/byoc.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wangch079) [reply connectors/byoc.py:380]: Yes, I did test it locally. I actually also added a test case in this PR:

https://github.com/elastic/connectors-python/blob/d7053db73522da993bd4d32e074af4012030c13f/connectors/tests/test_byoc.py#L912-L915

The change I have made in this line actually changed the behavior.

I the above example, it returns the same `filtering`, while it returns `{"advanced_snippet": {}, "rules": []}` before. That's why I want to confirm.

File: connectors/byoc.py (hunk 1 above)
Comment (wangch079) [reply connectors/byoc.py:267]: I kind of think this is part of our protocol, although it's not explicitly stated, it's just a convention.

In another PR, the draft rules will overwrite active rules, if the state is `VALID` after validation. I don't think we need to make it complicated.

File: connectors/byoc.py (hunk 4 above)
Comment (wangch079) [reply connectors/byoc.py:286]: `error` is used by another method:

https://github.com/elastic/connectors-python/blob/d7053db73522da993bd4d32e074af4012030c13f/connectors/byoc.py#L208-L210

It's not required to use the same method name as the job status.

File: connectors/byoc.py (hunk 3 above)
Comment (timgrein) [reply connectors/byoc.py:233]: Yes, that would be a problem. Also the new sync rule object model will probably use multiple filters for different tables. Therefore it definitely should be an array.

File: connectors/byoc.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (timgrein) [reply connectors/byoc.py:380]: Just double-checked as we're now passing an instance of `Filter` to the sync job, which already does the `value` extraction it should be fine to remove the explicit extraction within `SyncJob`.

File: connectors/byoc.py (hunk 4 above)
Comment (timgrein) [reply connectors/byoc.py:286]: Just thought it would be a bit cleaner/easier to read. But totally fine to keep it like that 👍

File: connectors/byoc.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wangch079) [reply connectors/byoc.py:380]: I'm not sure why the transformation is necessary:

In `connector` doc:
```json
{
  "advanced_snippet": {
     "value": {
       "query": {}
      }
   }, 
   "rules": []
}
```

which will be transformed to
```json
{
  "advanced_snippet": {
    "query": {}
   }, 
   "rules": []
}
```

Can we make it consistent before GA? cc @tarekziade

File: connectors/byoc.py [hunk 6]
```diff
@@ -178,96 +178,142 @@ async def all_connectors(self):
...
+        for key in keys:
+            if not isinstance(value, dict):
+                return default
+            value = value.get(key)
+        if value is None:
+            return default
+        return value
 
+    async def reload(self):
+        return await self.index.fetch_by_id(self.id)
```
Comment (tarekziade) [inline connectors/byoc.py:200]: what happens if `self.id` is `None` (it's possible with the current constructor)

scenario:

```python
doc = ESDocument('index')
await doc.reload()  
```

we might want to add a safeguard

File: connectors/byoc.py [hunk 7]
```diff
@@ -178,96 +178,138 @@ async def all_connectors(self):
...
+        return value
+
+    async def reload(self):
+        return await self.index.fetch_by_id(self.id)
+
+
+class SyncJob(ESDocument):
+    @property
+    def status(self):
+        return str2e(self.get("status"), JobStatus)
```
Comment (tarekziade) [inline connectors/byoc.py:206]: The default value needs to be an instance of `JobStatus(1)` or `JobStatus.PENDING` etc..

```python
>>> from connectors.byoc import JobStatus
>>> JobStatus
<enum 'JobStatus'>
>>> JobStatus(1)
<JobStatus.PENDING: 1>
```

File: connectors/byoc.py [hunk 8]
```diff
@@ -178,96 +178,138 @@ async def all_connectors(self):
...
 
-class SyncJob:
-    def __init__(self, elastic_index, connector_id, doc_source=None):
-        self.connector_id = connector_id
-        self.elastic_index = elastic_index
-        self.created_at = datetime.now(timezone.utc)
-        self.completed_at = None
-        self.filtering = Filter()
-        self.client = elastic_index.client
+class ESDocument:
```
Comment (tarekziade) [inline connectors/byoc.py:181]: Should we move that class to `connectors/es/document.py` ?

File: connectors/byoc.py (hunk 7 above)
Comment (tarekziade) [reply connectors/byoc.py:206]: oh sorry, I just realized how the  `str2e` works by grabbing the `Enum` class.

Forget about this comment. On a side note, this is very hard to understand and not semantically correct. We should make this simpler later.

I think we'll want this to work eventually:

```
@property
def status(self):
    return JobStatus(self.get("status"))
```

**Comment by tarekziade** [comment]
related https://github.com/elastic/connectors-python/pull/464

File: connectors/byoc.py (hunk 6 above)
Comment (wangch079) [reply connectors/byoc.py:200]: What do you think we add to `__init__` of `ESDocument`?

```python
class InvalidDocumentSourceError(Exception):
    pass


class ESDocument:
    def __init__(self, elastic_index, doc_source=None):
        self.index = elastic_index
        if not isinstance(doc_source, dict):
            raise InvalidDocumentSourceError(f"Invalid doc_source type, expected: {dict.__name__}, got: {doc_source.__class__.__name__}")
        self.id = doc_source.get("_id")
        if not isinstance(self.id, str):
            raise InvalidDocumentSourceError(
                f"Invalid id type, expected: {str.__name__}, got: {self.id.__class__.__name__}")
        self._source = doc_source.get("_source", {})
        if not isinstance(self._source, dict):
            raise InvalidDocumentSourceError(
                f"Invalid source type, expected: {dict.__name__}, got: {self._source.__class__.__name__}")
```

File: connectors/byoc.py (hunk 7 above)
Comment (wangch079) [reply connectors/byoc.py:206]: Yes, we discussed this before, this will work:
```python
class JobStatus(Enum):
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    CANCELING = "canceling"
    CANCELED = "canceled"
    SUSPENDED = "suspended"
    COMPLETED = "completed"
    ERROR = "error"

# enum to str
JobStatus.PENDING.value

# str to enum
JobStatus("pending")
```

But we also need to consider the edge case when it's `None`, probably as you suggested before, we could have a special `unset` value for each enum.

File: connectors/byoc.py (hunk 6 above)
Comment (wangch079) [reply connectors/byoc.py]: Eventually I didn't use `isinstance(doc_source, dict)` as the response from the ES call is an `ObjectApiResponse` object, not a `dict`. Here is the change: https://github.com/elastic/connectors-python/pull/460/commits/5d03016809dcc71b19b10fb38df21dbfd79efa46

File: connectors/byoc.py (hunk 8 above)
Comment (wangch079) [reply connectors/byoc.py]: Done in https://github.com/elastic/connectors-python/commit/5d03016809dcc71b19b10fb38df21dbfd79efa46

File: connectors/byoc.py (hunk 7 above)
Comment (wangch079) [reply connectors/byoc.py:184]: Consolidated the proposal in the comment: https://github.com/elastic/connectors-python/pull/464#issuecomment-1429286586

File: connectors/byoc.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wangch079) [reply connectors/byoc.py]: Tech-tebt issue created: https://github.com/elastic/enterprise-search-team/issues/3950

[the remark this answers is not in the document]
Comment (timgrein) [reply connectors/byoc.py]: Thank you for creating the issue 👍

File: connectors/es/document.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+#
+
+from connectors.logger import logger
+
+
+class InvalidDocumentSourceError(Exception):
+    pass
+
+
+class ESDocument:
```
Comment (tarekziade) [inline connectors/es/document.py:14]: This block can be greatly simplified like this:

```python
class ESDocument:
    def __init__(self, elastic_index, doc_source):
        self.index = elastic_index
        self.id = doc_source.get("_id")
        if not isinstance(self.id, str):
            raise InvalidDocumentSourceError(f"Invalid format found for id: {self.id}")

        self._source = doc_source.get("_source", {})
        if not isinstance(self._source, dict):
            raise InvalidDocumentSourceError(
                f"Invalid format found for source: {self._source}"
            )
```

Comment (wangch079) [reply connectors/es/document.py:14]: The reason I wrap it with `try... except`, is to handle edge case where some funny value is passed in for `doc_source`, e.g. `None`, or a string like `"foobar"`

File: connectors/tests/test_es_document.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        ({}, InvalidDocumentSourceError),
+        ({"_id": {}}, InvalidDocumentSourceError),
+        ({"_id": "1", "_source": "hahaha"}, InvalidDocumentSourceError),
+        ({"_id": "1", "_source": {}}, None),
+    ],
+)
+def test_es_document(doc_source, exception, patch_logger):
+    try:
+        ESDocument(elastic_index=None, doc_source=doc_source)
+    except Exception as e:
```
Comment (tarekziade) [inline connectors/tests/test_es_document.py:25]: look at `pytest.raises` https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions

File: connectors/byoc.py [hunk 11]
```diff
@@ -181,96 +181,116 @@ async def all_connectors(self):
...
+    async def claim(self):
+        doc = {
+            "status": JobStatus.IN_PROGRESS.value,
+            "started_at": iso_utc(),
+            "last_seen": iso_utc(),
+            "worker_hostname": socket.gethostname(),
+        }
+        await self.index.update(doc_id=self.id, doc=doc)
+
+    async def done(self, ingestion_stats={}, connector_metadata={}):
```
Comment (tarekziade) [inline connectors/byoc.py:258]: You can't have this signature:
```
async def done(self, ingestion_stats={}, connector_metadata={}):
```

do:

```
async def done(self, ingestion_stats=None, connector_metadata=None):
    if ingestion_stats is None:
        ingestion_stats = {}
```

reason: https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

File: connectors/es/document.py (hunk 9 above)
Comment (wangch079) [reply connectors/es/document.py:14]: This is exactly what I proposed initially: https://github.com/elastic/connectors-python/pull/460#discussion_r1105347125, and why I didn't go that way: https://github.com/elastic/connectors-python/pull/460#discussion_r1105420783

I'm also not trying to check `isinstance(doc_source, ObjectApiResponse)`, as I think if a `dict` is provided as the `doc_source`, it should work as well.

File: connectors/tests/test_es_document.py (hunk 10 above)
Comment (wangch079) [reply connectors/tests/test_es_document.py:25]: Yes, I planned to use `pytest.raises` initially, but there are test cases which don't raise errors, and `pytest.raises` won't work.

Then I say this: https://miguendes.me/how-to-check-if-an-exception-is-raised-or-not-with-pytest#heading-how-to-assert-that-no-exception-is-raised, and I think this pattern works well for this particular test.

File: connectors/byoc.py (hunk 11 above)
Comment (wangch079) [reply connectors/byoc.py:258]: TIL

Comment (wangch079) [reply connectors/byoc.py]: Done in https://github.com/elastic/connectors-python/pull/460/commits/25427141df45140874ae859b947218f2c950d726

File: connectors/es/document.py (hunk 9 above)
Comment (tarekziade) [reply connectors/es/document.py:14]: you should still try to be explicit because having different types allowed for  `doc_source` makes the instanciation of this class a bit fuzzy.

If you want to support several types as inputs, what is usually done is the following:

```
class ESDocument:
    def __init__(self, elastic_index, doc_source):
        self.index = elastic_index

        if not isinstance(doc_source, dict):            
            raise InvalidDocumentSourceError(f"Invalid format found for doc_source: {str(doc_source)}")
 
        self.id = doc_source.get("_id")
        if not isinstance(self.id, str):
            raise InvalidDocumentSourceError(f"Invalid format found for id: {self.id}")

        self._source = doc_source.get("_source", {})
        if not isinstance(self._source, dict):
            raise InvalidDocumentSourceError(
                f"Invalid format found for source: {self._source}"
            )

    @classmethod 
    def from_api_response(cls, api_response):
        return cls(api_response.body)    
```

File: connectors/tests/test_es_document.py (hunk 10 above)
Comment (tarekziade) [reply connectors/tests/test_es_document.py:25]: > there are test cases which don't raise errors, and pytest.raises won't work.

if you don't want to assert that an error was not raised, don't use `pytest.raises` for sure, the problem is your attempt to parameterized even when it's not raising..

here's your test fixed (untested):

```
@pytest.mark.parametrize(
    "doc_source, exception",
    [None,  "hahaha", {}, {"_id": {}}, {"_id": "1", "_source": "hahaha"}] 
)
def test_es_document_raise(doc_source, patch_logger):
    with pytest.raises(InvalidDocumentSourceError):
        ESDocument(elastic_index=None, doc_source=doc_source)

def test_es_document_ok():
    doc = ESDocument(elastic_index=None, doc_source={"_id": "1", "_source": {}})
    assert something on doc
```

File: connectors/es/document.py (hunk 9 above)
Comment (wangch079) [reply connectors/es/document.py:14]: `ObjectApiResponse` is defined in `elastic-transport`, and has method `__getitem__` defined:
https://github.com/elastic/elastic-transport-python/blob/6d6e64ae26a4088801f6df1f3bdb40fea88b9b75/elastic_transport/_response.py#L177-L188

The reason why I don't want to use `isinstance(doc_source, ObjectApiResponse)`:
1. It should work for `dict` as well (actually any object with method `get`)
2. I don't want to add dependency on `elastic_transport` here

Or we could simply do this: https://github.com/elastic/connectors-python/pull/460#discussion_r1106263366, which will raise any possible exception, instead of `InvalidDocumentSourceError`

File: connectors/tests/test_es_document.py (hunk 10 above)
Comment (wangch079) [reply connectors/tests/test_es_document.py]: Done here: https://github.com/elastic/connectors-python/pull/460/commits/b2b1060798ce2f2e401951b609d49870c09fd5e2

File: connectors/es/document.py (hunk 9 above)
Comment (tarekziade) [reply connectors/es/document.py:14]: > It should work for dict as well (actually any object with method get)

Can you elaborate on this?  Do you mean it should support the mapping interface? Will we really use this class for any kind of object ? 

> I don't want to add dependency on elastic_transport here

In that case you can check for `get` presence:

```python
if not hasattr(data_source, 'get'):
    raise InvalidDocumentSourceError("Expected a class with get")
```

But it would be best to be explicit about what the class expects IMHO. 
You could also make it compatible only with dict and ask the caller to take care of passing a dict, so you separate concerns. THe code that manipulates `ObjectApiResponse` could simply do:

```
ESDocument(index, my_response.body)
```

File: connectors/tests/test_sync.py [hunk 12]
```diff
@@ -700,5 +730,3 @@ async def test_concurrent_syncs(mock_responses, patch_logger, set_env):
 
     # make sure we synced the three connectors
     patch_logger.assert_present("[1] Sync done: 1 indexed, 0  deleted. (0 seconds)")
-    patch_logger.assert_present("[2] Sync done: 1 indexed, 0  deleted. (0 seconds)")
```
Comment (tarekziade) [inline connectors/tests/test_sync.py:703]: why do you remove this coverage? we want to make sure we sync the 3  connectors

Comment (wangch079) [reply connectors/tests/test_sync.py:703]: Because I will soon rewrite this test when I refactor `SyncService`, where I will mock `ConcurrentTasks`, and 
```python
calls = [
    call(functools.partial(self._one_sync, connector1, es)),
    call(functools.partial(self._one_sync, connector2, es)),
    functools.partial(self._one_sync, connector3, es),
]
syncs.put.assert_has_awaits(calls)
```

I will fix the test in this PR.

File: connectors/es/document.py (hunk 9 above)
Comment (wangch079) [reply connectors/es/document.py:14]: >Can you elaborate on this?

If a caller wants to instantiate an `ESDocument` object, he/she can pass in an `ObjectApiResponse` object (as what we do in `connectors-python`:
```python
resp = a_client.get(index=".elastic-connectors`, id="1") # an ObjectApiResponse object
doc = ESDocument(an_index, resp)
```
he/she can also pass in a dict:
```python
connector_doc = {
    "_id": "1",
    "_source": {}
}
doc = ESDocument(an_index, connector_doc)
```

Or any object with `get` method.

But again, `YAGNI`.

The reason why I don't want to handle `ObjectApiResponse` explicitly, it's that I'm hesitate to have `ObjectApiResponse` imported as a dependency:
```python
# _response module starts with _ implies it's not supposed to be used directly?
from elastic_transport._response import ObjectApiResponse

class ESDocument:
    def __init__(self, elastic_index, doc_source):
        self.index = elastic_index
        if not isinstance(doc_source, ObjectApiResponse):
            raise InvalidDocumentSourceError(
                f"Invalid doc_source type found: {doc_source.__class__.__name__}, expected: {ObjectApiResponse.__class__.__name__}"
            )
        ...
```

Comment (tarekziade) [reply connectors/es/document.py:14]: But what is the intent of `ESDocument` -- can you add a two-sentence description in its docstring?

Why a `dict` or an `ObjectApiResponse`?
Will it ever be instantiated with something else than a document from an Elasticsearch client? 
It seems to me that we will never use that class with different type of data.

In any case, if you want both, you don't need to import `elastic_transport`, what prevents you to do: 

```
resp = a_client.get(index=".elastic-connectors`, id="1") # an ObjectApiResponse object
doc = ESDocument(an_index, resp.body)
```
Because your class just needs the dict from the body and that's what `ObjectApiResponse.body` returns

File: connectors/tests/test_sync.py (hunk 12 above)
Comment (tarekziade) [reply connectors/tests/test_sync.py:703]: ah ok, great thanks! let's make sure we don't lose that coverage

File: connectors/es/document.py (hunk 9 above)
Comment (wangch079) [reply connectors/es/document.py:14]: >Will it ever be instantiated with something else than a document from an Elasticsearch client?

No, not for now.

> In any case, if you want both, you don't need to import `elastic_transport`, what prevents you to do

Will update.

Comment (wangch079) [reply connectors/es/document.py:10]: Done here: https://github.com/elastic/connectors-python/pull/460/commits/f7bb6fd9bb919caaad753682cb8335172a60b55e

File: connectors/tests/test_sync.py (hunk 12 above)
Comment (wangch079) [reply connectors/tests/test_sync.py]: Done here: https://github.com/elastic/connectors-python/pull/460/commits/73454007a23342b0b04346d5f2230da9351b6942

**Review by tarekziade — APPROVED** [review]
Thanks for all the interesting discussions and bearing with all my comments

**Comment by wangch079** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.7|[](https://github.com/elastic/connectors-python/pull/504)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)
